pkg/controller/pod_volume_restore_controller.go

```diff
@@ -250,40 +252,99 @@ func (c *podVolumeRestoreController) processRestore(req *arkv1api.PodVolumeResto
+	// restoring, you can't restore to /baz/volume. You may restore to /baz/foo/bar/volume, though. The net result of
+	// all this is that we can't restore directly into the new volume's directory, because the path is entirely different
+	// than the backed-up one.
 	if stdout, stderr, err = runCommand(resticCmd.Cmd()); err != nil {
-		log.WithError(errors.WithStack(err)).Errorf("Error running command=%s, stdout=%s, stderr=%s", resticCmd.String(), stdout, stderr)
-		return c.fail(req, fmt.Sprintf("error running restic restore, stderr=%s: %s", stderr, err.Error()), log)
+		return errors.Wrapf(err, "error running restic restore, cmd=%s, stdout=%s, stderr=%s", resticCmd.String(), stdout, stderr)
 	}
 	log.Debugf("Ran command=%s, stdout=%s, stderr=%s", resticCmd.String(), stdout, stderr)
 
+	// Now, get the full path of the restored volume in the staging directory, which will
+	// look like:
+	// 		/restores/<new-pod-uid>/host_pods/<backed-up-pod-uid>/volumes/<volume-plugin-name>/<volume-dir>
+	restorePath, err := singlePathMatch(fmt.Sprintf("/restores/%s/host_pods/*/volumes/*/%s", string(req.Spec.Pod.UID), volumeDir))
+	if err != nil {
+		return errors.Wrap(err, "error identifying path of restore staging directory")
+	}
+
+	// Also get the full path of the new volume's directory (as mounted in the daemonset pod), which
+	// will look like:
+	// 		/host_pods/<new-pod-uid>/volumes/<volume-plugin-name>/<volume-dir>
+	volumePath, err := singlePathMatch(fmt.Sprintf("/host_pods/%s/volumes/*/%s", string(req.Spec.Pod.UID), volumeDir))
+	if err != nil {
+		return errors.Wrap(err, "error identifying path of volume")
+	}
+
+	// Move the contents of the staging directory into the new volume directory to finalize the restore. This
+	// is being executed through a shell because attempting to do the same thing in go (via os.Rename()) is
+	// giving errors about renames not being allowed across filesystem layers in a container. This is occurring
+	// whether /restores is part of the writeable container layer, or is an emptyDir volume mount. This may
+	// be solvable but using the shell works so not investigating further.
+	//
+	// Glob patterns:
+	// 	[^.]*  : any non-dot character followed by anything (regular files)
+	// 	.[^.]* : a dot followed by a non-dot and anything else (dotfiles starting with a single dot, excluding '.')
+	// 	..?*   : two dots followed by any single character and anything else (dotfiles starting with two dots, excluding '..')
+	if err := moveMatchingFiles(restorePath, volumePath, "[^.]*", ".[^.]*", "..?*"); err != nil {
+		return errors.Wrapf(err, "error moving files from restore staging directory into volume")
+	}
+
+	// The staging directory should be empty at this point since we moved everything, but
+	// make sure.
+	if err := os.RemoveAll(restorePath); err != nil {
+		return errors.Wrap(err, "error removing restore staging directory and its contents")
```

Comment:
Are there any conceivable situations where we successfully moved the matching files, but we get an error on the RemoveAll call? Do we need to be more detailed here in reporting the error - list all the files/dirs in `restorePath` or something?

Reply:
If we return here, we never write the done file, so the pod's init container will be waiting forever. What do you think we can do to shed more light for these waiting init containers?

Reply:
I suppose it is possible there's an error. Now that you raise it, I'm thinking this shouldn't be an error for the restore - since this is just (probably redundant) cleanup in the daemonset pod.

Reply:
I think it makes sense just to add a warning to the server log here - since this should just be removing the empty staging directory. Do you think that's reasonable?

Reply:
Updated.